Added Elevation to Motorcycle Routing

boldtrn opened the pull request:
Basically I added elevation data to the motorcycle routing. I worked with srtm elevation data. The values significantly changed when using the cigar data. But somehow the cigar data seemed not to work that good? Even though you mention it would be the better one.

The elevation widget for the cigar data didn't work in the UI.

Fixed a serious bug when calculating the bendiness. Results should be better now.

BTW: Why did you changed the min bendiness in the CurvatureWeighting to 1? Actually the Bendiness can be 0.
BTW2: Successfully squashed the commits ;) 

Best,
Robin

karussell commented:
> But somehow the cigar data seemed not to work that good? Even though you mention it would be the better one.

Yes, they are better. What do you mean with 'not working' in this context?

> BTW: Why did you changed the min bendiness in the CurvatureWeighting to 1? Actually the Bendiness can be 0

Sorry if I 'fixed' something without asking you for assistance. This values needs to be exact (it should not overestimate) and I tried to understand all the max and min values of the formula you used and obviously made some mistakes. I'll have a look again.

karussell commented:
Would you mind to fix the formatting that e.g. the opening bracket has to be on a new line? E.g. use NetBeans or the IntelliJ settings in the core/files folder or manually adjust Eclipse. Otherwise there are many unimportant changes here.

boldtrn commented:
Allright formating should be fixed.
Also minBendiness is at .1. Shouldn't become lower than this.

You'r welcome ;)

> Yes, they are better. What do you mean with 'not working' in this context?

It does not show anything, at least on my machine, not sure if this is something local. But haven't tried in depth.

boldtrn commented:
Ok, I just changed some things. I adjusted the min length that a edge needs to have to be even calculated and the minimum number of nodes (3).

Also I did a tag based rating. Streets that have sidewalks or lighting are rated worse.

I changed the speeds of the streets. Especially the speed of secondary and tertiary. You can drive there way faster on a motorcycle.

karussell commented on core/src/main/java/com/graphhopper/routing/util/MotorcycleFlagEncoder.java:
```diff
@@ -350,19 +365,63 @@ public void applyWayTags( OSMWay way, EdgeIteratorState edge )
         double roadDistance = edge.getDistance();
-        double beelineDistance = getBeelineDistance(way);
+        double beelineDistance = getBeelineDistance(edge);
+        PointList pointList = edge.fetchWayGeometry(3);
```
I would reduce the number of calling the relative expensive fetchWayGeometry. E.g. pass the pointList to the getBeelineDistance method?

boldtrn replied:
Yes sure, I thought about that as well, but thought the method call would not be expensive.

karussell commented on core/src/test/java/com/graphhopper/routing/RoutingAlgorithmIT.java:
```diff
@@ -104,13 +105,14 @@ public void testMonacoMotorcycle()
-        list.add(new OneRun(43.733802, 7.413433, 43.739662, 7.424355, 2423, 141));
-        list.add(new OneRun(43.730949, 7.412338, 43.739643, 7.424542, 2253, 120));
+        list.add(new OneRun(43.733802, 7.413433, 43.739662, 7.424355, 5636, 298));
+        list.add(new OneRun(43.730949, 7.412338, 43.739643, 7.424542, 5467, 277));
```
Did you look if the new routes in monaco are more reasonable or why is there this big difference?

boldtrn replied:
To be honest, no I haven't checked them. Something quite crazy is happening here. With this file `core/files/monaco.osm.gz` the results are crazy (and actually not reasonable) and the distances are as set above. 
If I use the osm.pbf file from http://download.geofabrik.de/europe/monaco.html, the distances are fine and the routes are absolutely ok. So I really don't know what to do here?

karussell replied:
Keep in mind that "shortest" is used and a relative old version of monaco ... but I'll have a look if you don't find the reason. Maybe due to the elevation?

boldtrn replied:
Yes I compared it with shortest. I think the issue is related to speed limites. Probably in the old version of Monaco the speed limits are not set properly. Since I changed the default speed lmits, the routing could be wrong. That would be my guess, since the elevation thing is only relevant when using the curvature routing.

karussell commented:
Thanks for the improvements! Would you revert the formatting changes in OSMReader and the tests? Also graphhopper.sh needs a revert so that we'll fix this in the separate issue.

karussell commented on core/src/main/java/com/graphhopper/routing/util/MotorcycleFlagEncoder.java:
```diff
@@ -93,30 +100,37 @@ public MotorcycleFlagEncoder( int speedBits, double speedFactor, int maxTurnCost
         // forestry stuff
         defaultSpeedMap.put("track", 15);
 
+        enjoyableTags.add("scenic");
```
Interesting! Did not knew about this. Is this usable for bike too @ratrun ?

http://taginfo.openstreetmap.org/keys/scenic

boldtrn replied:
Unfortunately it is not used very often, especially on roads. It's oftimes used on trails and paths http://overpass-turbo.eu/s/cVF. But there are some roads and according to the wiki it should be used for beautiful roads e.g. http://wiki.openstreetmap.org/wiki/Talk:Key:highway#Scenic.2Fhistorical_roads.

ratrun replied:
Thank you for the hint about the scenic tag. I didn't know it, although I did search for something like this a long time ago, The wiki page was created in 2012 and I searched before that time. I'll prepare a PR for the bicycles. I'll intent to increase the preferene by one level. Although this changes looks easy it might take some time.

karussell replied:
nice and no time pressure please :) !

karussell commented on core/src/main/java/com/graphhopper/routing/util/MotorcycleFlagEncoder.java:
```diff
@@ -93,30 +100,37 @@ public MotorcycleFlagEncoder( int speedBits, double speedFactor, int maxTurnCost
+        enjoyableTags.add("scenic");
+
+        // List is typcially set in urban areas
+        frustratingTags.add("lit");
```
Oh, why is this frustrating :) ?

boldtrn replied:
Exactly that was a quick and probably dirty fix, but the results work quite well, at least for my area. But you are right if #230 is fixed we should use these results. I hope you are fine if we leave this here until #230 is ready?

But in general lit roads are a sign of high traffic volume and we don't want traffic.

karussell replied:
Sure, this is fine. BTW: also #75 could be used to determine slow down.

boldtrn commented:
Your comments except for the tests are pushed. I can squash them when its mergeable.
